File: python/Remove Nodes From Linked List.py
# Define the ListNode class
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# Function to reverse a linked list
def reverse_list(head):
    prev = None
    current = head
    next_temp = None

    # Set each node's next pointer to the previous node
    while current:
        next_temp = current.next
        current.next = prev
        prev = current
        current = next_temp

    return prev


# Function to remove nodes
def removeNodes(head):
    # Reverse the original linked list
    head = reverse_list(head)

    maximum = 0
    prev = None
    current = head

    # Traverse the list deleting nodes
    while current:
        maximum = max(maximum, current.val)

        # Delete nodes that are smaller than maximum
        if current.val < maximum:
            # Delete current by skipping
            if prev:
                prev.next = current.next
            deleted = current
            current = current.next
            deleted.next = None

        # Current does not need to be deleted
        else:
            prev = current
            current = current.next

    # Reverse and return the modified linked list
    return reverse_list(head)


# Function to convert a list to a linked list
def list_to_linked_list(lst):
    if not lst:
        return None
    head = ListNode(lst[0])
    current = head
    for val in lst[1:]:
        current.next = ListNode(val)
        current = current.next
    return head


# Function to print the linked list
def print_linked_list(head):
    current = head
    while current:
        print(current.val, end=" -> ")
        current = current.next
    print("None")


# Example usage:
input_list = [1,1,1,1]
head = list_to_linked_list(input_list)

result = removeNodes(head)

# Print the result
print_linked_list(result)
'''
You are given the head of a linked list.

Remove every node which has a node with a greater value anywhere to the right side of it.

Return the head of the modified linked list.

 

Example 1:


Input: head = [5,2,13,3,8]
Output: [13,8]
Explanation: The nodes that should be removed are 5, 2 and 3.
- Node 13 is to the right of node 5.
- Node 13 is to the right of node 2.
- Node 8 is to the right of node 3.
Example 2:

Input: head = [1,1,1,1]
Output: [1,1,1,1]
Explanation: Every node has value 1, so no nodes are removed.
 

Constraints:

The number of the nodes in the given list is in the range [1, 105].
1 <= Node.val <= 105'''

File: JAVA/java/src/RemoveNodesFromLinkedList.java
class ListNode13 {
    int val;
    ListNode13 next;
    ListNode13() {}
    ListNode13(int val) { this.val = val; }
    ListNode13(int val, ListNode13 next) { this.val = val; this.next = next; }
}

public class RemoveNodesFromLinkedList {
    public ListNode13 reverseList(ListNode13 head) {
        ListNode13 prev = null;
        ListNode13 current = head;
        ListNode13 nextTemp = null;

        // Set each node's next pointer to the previous node
        while (current != null) {
            nextTemp = current.next;
            current.next = prev;
            prev = current;
            current = nextTemp;
        }

        return prev;
    }

    public ListNode13 removeNodes(ListNode13 head) {
        // Reverse the original linked list
        head = reverseList(head);

        int maximum = 0;
        ListNode13 prev = null;
        ListNode13 current = head;

        // Traverse the list deleting nodes
        while (current != null) {
            maximum = Math.max(maximum, current.val);

            // Delete nodes that are smaller than maximum
            if (current.val < maximum) {
                // Delete current by skipping
                if (prev != null) {
                    prev.next = current.next;
                }
                ListNode13 deleted = current;
                current = current.next;
                deleted.next = null;
            }

            // Current does not need to be deleted
            else {
                prev = current;
                current = current.next;
            }
        }

        // Reverse and return the modified linked list
        return reverseList(head);
    }

    // Function to convert an array to a linked list
    public ListNode13 arrayToLinkedList(int[] arr) {
        if (arr == null || arr.length == 0) {
            return null;
        }
        ListNode13 head = new ListNode13(arr[0]);
        ListNode13 current = head;
        for (int i = 1; i < arr.length; i++) {
            current.next = new ListNode13(arr[i]);
            current = current.next;
        }
        return head;
    }

    // Function to print the linked list
    public void printLinkedList(ListNode13 head) {
        ListNode13 current = head;
        while (current != null) {
            System.out.print(current.val + " -> ");
            current = current.next;
        }
        System.out.println("null");
    }

    // Example usage:
    public static void main(String[] args) {
        RemoveNodesFromLinkedList solution = new RemoveNodesFromLinkedList();
        int[] inputArray = {5, 2, 13, 3, 8};
        ListNode13 head = solution.arrayToLinkedList(inputArray);
        ListNode13 result = solution.removeNodes(head);
        solution.printLinkedList(result);
    }
}
